examples/usercall-extension-bind/runner/src/main.rs

```diff
@@ -0,0 +1,356 @@
+    fn read(&self, rdr: &mut impl Read) -> IoResult<(Option<SocketAddr>, Option<SocketAddr>)> {
+       match self {
+            ProxyAddrReader::V4(len) => {
+                if *len as usize != size_of::<ProxyIpV4Addr>() {
+                    Err(Error::new(ErrorKind::InvalidData, "Unexpected address length received"))?;
+                }
+                let addr = ProxyIpV4Addr::from_reader(rdr)?;
+                let (local, peer) = addr.into();
+                Ok((Some(local), Some(peer)))
+            },
+            ProxyAddrReader::V6(len) => {
+                if *len as usize != size_of::<ProxyIpV6Addr>() {
+                    Err(Error::new(ErrorKind::InvalidData, "Unexpected address length received"))?;
+                }
+                let addr = ProxyIpV6Addr::from_reader(rdr)?;
+                let (local, peer) = addr.into();
+                Ok((Some(local), Some(peer)))
+            },
+            ProxyAddrReader::UNSPEC(len) => {
+                       let mut buf = vec![0; *len as _];
+                       let _ = rdr.read_exact(&mut buf[..])?;
+                       Ok((None, None))
+                      },
+        }
+    }
+}
+
+#[allow(unused)]
+struct ProxyHdrV2 {
+    sig     : [u8; 12], /* hex 0D 0A 0D 0A 00 0D 0A 51 55 49 54 0A */
+    ver_cmd : u8,       /* protocol version and command*/
+    fam     : u8,       /* protocol family and address*/
+    len     : u16,      /* number of following bytes part of the header*/
+    addr_reader : ProxyAddrReader /* read len bytes into a tuple of SocketAddr */
+}
+
+impl ProxyHdrV2 {
+    fn from_reader(rdr: &mut impl Read) -> IoResult<Self> {
+        let mut sig : [u8; 12] = [0; 12];
+        let _ = rdr.read_exact(&mut sig[..])?;
+        if &sig[..] != HEADER_SIG {
+            return Err(Error::new(ErrorKind::InvalidData, "Protocol header signature mismatch"))
+        }
+        let ver_cmd = rdr.read_u8()?;
+        let fam = rdr.read_u8()?;
+        let len = rdr.read_u16::<NetworkEndian>()?;
+        let addr_reader = match (ver_cmd, fam) {
+            (V2CMD_LOCAL, FAMILY_UNSPEC) => ProxyAddrReader::UNSPEC(len),
+            (V2CMD_PROXY, FAMILY_TCP4) =>   ProxyAddrReader::V4(len),
+            (V2CMD_PROXY, FAMILY_TCP6) =>   ProxyAddrReader::V6(len),
+            _ => Err(Error::new(ErrorKind::InvalidData, "Unsupported version/command/family"))?
+        };
+    
+        Ok(ProxyHdrV2 {
+               sig,
+               ver_cmd,
+               fam,
+               len,
+               addr_reader,
+          })
+    }
+}
+fn read_proxy_protocol_header(stream :&mut TcpStream) -> IoResult<(
+                                                                  Option<SocketAddr>,
+                                                                  Option<SocketAddr>)> {
+
+    let hdr = ProxyHdrV2::from_reader(stream)?;
+    hdr.addr_reader.read(stream)
+}
+
+
+struct ProxyProtocol {
+    listener      : TcpListener,
+    pub local_address : String,
```

Comment:
Does this need to be stored? Can you just return this from `fn new`?

Reply:
Incorporated in recent push.